EventGenerator/src/FlatEMinusGun_module.cc

```diff
@@ -0,0 +1,152 @@
+    explicit Flateminus(const Parameters& conf);
+
+    virtual void produce(art::Event& event) override;
+
+    //----------------------------------------------------------------
+  private:
+    Float_t pmag_gen;
+    Float_t time_gen;
+    Float_t x_gen;
+    Float_t y_gen;
+    Float_t z_gen;
+    TTree*  _Ntup;
+    const PDGCode::type electronId_ = PDGCode::e_minus;
+    double electronMass_;
+    double startMom_;
+    double endMom_;
+    double muonLifeTime_;
+
+    art::ProductToken<SimParticleCollection> const simsToken_;
+
+    unsigned verbosity_;
+
+    art::RandomNumberGenerator::base_engine_t& eng_;
+    CLHEP::RandFlat randFlat_;
+    CLHEP::RandExponential randExp_;
+    RandomUnitSphere   randomUnitSphere_;
+    bool makeHistograms_;
+  };
+
+  //================================================================
+  Flateminus::Flateminus(const Parameters& conf)
+    : EDProducer{conf}
+    , electronMass_(GlobalConstantsHandle<ParticleDataTable>()->particle(electronId_).ref().mass().value())
+    , startMom_(conf().startMom())
+    , endMom_(conf().endMom())
+    , muonLifeTime_{GlobalConstantsHandle<PhysicsParams>()->getDecayTime(conf().stoppingTargetMaterial())}
+    , simsToken_{consumes<SimParticleCollection>(conf().inputSimParticles())}
+    , verbosity_{conf().verbosity()}
+    , eng_{createEngine(art::ServiceHandle<SeedService>()->getSeed())}
+    , randFlat_{eng_}
+    , randExp_{eng_}
+    , randomUnitSphere_{eng_}
+    , makeHistograms_(conf().makeHistograms())
+  {
+    produces<mu2e::StageParticleCollection>();
+
+
+    if(makeHistograms_){
+      art::ServiceHandle<art::TFileService> tfs;
+      art::TFileDirectory tfdir = tfs->mkdir( "IPAGun" );
+      _Ntup  = tfs->make<TTree>("GenTree", "GenTree");
+	    _Ntup->Branch("pmag_gen", &pmag_gen , "pmag_gen/F");
+	    _Ntup->Branch("time_gen", &time_gen, "time_gen/F");
+	    _Ntup->Branch("x_gen", &x_gen, "x_gen/F");
+	    _Ntup->Branch("y_gen", &y_gen, "y_gen/F");
+	    _Ntup->Branch("z_gen", &z_gen, "z_gen/F");
+	    }
+  }
+
+  //================================================================
+  void Flateminus::produce(art::Event& event) {
+    auto output{std::make_unique<StageParticleCollection>()};
+
+    const auto simh = event.getValidHandle<SimParticleCollection>(simsToken_);
+    const auto mus = stoppedMuMinusList(simh);
+
+    if(mus.empty()) {
+      throw   cet::exception("BADINPUT")
+        <<"Flateminus::produce(): no suitable stopped mu- in the input SimParticleCollection\n";
+
+    }
+    
+    const auto mustop = mus.at(eng_.operator unsigned int() % mus.size());
+    double randomMom = randFlat_.fire(startMom_, endMom_);
+    double randomE = sqrt(electronMass_*electronMass_ + randomMom*randomMom);
+    double time = mustop->endGlobalTime() + randExp_.fire(muonLifeTime_);
+    output->emplace_back(mustop,
+                         ProcessCode::mu2eFlateMinus,
+                         PDGCode::e_minus,
```

Comment:
Looking ahead to when we need flat eplus, couldn't this same module be used to create that just by making the particle code a parameter?   That would reduce code duplication, since everything else in the module would be identical.  You would also have to set the processcode consistently at module construction time, and throw if there isn't an appropriate process code.

Reply:
True, I can add that funtionality

Reply:
Yes, the old way used that module

Reply:
For the process code and PDG - can we take these in as FCL parameters?